packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -0,0 +1,254 @@
+   */
+  constructor(bool test) public InitializableV2(test) {}
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _registry The address of the registry.
+   * @param _spread The spread charged on exchanges.
+   */
+  function initialize(address _registry, uint256 _spread) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(_registry);
+    setSpread(_spread);
+  }
+
+  /**
+   * @notice Creates a new exchange proposal and deposits the tokens being sold.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The proposal identifier for the newly created exchange proposal.
+   */
+  function createExchangeProposal(address stableToken, uint256 sellAmount, bool sellCelo)
+    external
+    nonReentrant
+    returns (uint256)
+  {
+    // Require the configurable stableToken max exchange amount to be > 0.
+    // This covers the case where a stableToken has never been explicitly permitted.
+    ExchangeLimits memory exchangeLimits = stableTokenExchangeLimits[stableToken];
+    require(exchangeLimits.maxExchangeAmount > 0, "Max stable token exchange amount must be > 0");
+
+    // Using the current oracle exchange rate, calculate what the buy amount is.
+    // This takes the spread into consideration.
+    uint256 buyAmount = getBuyAmount(stableToken, sellAmount, sellCelo);
+
+    // Ensure that the amount of stableToken being bought or sold is within
+    // the configurable exchange limits.
+    uint256 stableTokenExchangeAmount = sellCelo ? buyAmount : sellAmount;
+    require(
+      stableTokenExchangeAmount <= exchangeLimits.maxExchangeAmount &&
+        stableTokenExchangeAmount >= exchangeLimits.minExchangeAmount,
+      "Stable token exchange amount not within limits"
+    );
+
+    // Deposit the assets being sold.
+    IERC20 sellToken = sellCelo ? getGoldToken() : IERC20(stableToken);
+    require(
+      sellToken.transferFrom(msg.sender, address(this), sellAmount),
+      "Transfer of sell token failed"
+    );
+
+    // Record the proposal.
+    uint256 proposalId = exchangeProposalCount;
+    exchangeProposals[proposalId] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
+      buyAmount: buyAmount,
```

Comment:
shouldn't this be set in units in case of stableToken too?

Reply:
Another one where I don't have a super strong opinion but just for implementation simplicity I kept it in value. I go into more detail in the PR description, but storing the buy amount as a value means that in a CELO -> StableToken trade, the stable token value amount is calculated at the proposal time, and that same value is given to the exchanger when the exchange is eventually executed. Meaning that any inflation that would've been incurred in the time [proposal time, execution time] is not considered. This is more advantageous to the exchanger. Wdyt?